stdlib/asm/math/u64.masm

```diff
@@ -580,8 +718,58 @@ end
 # Stack transition looks as follows:
 # [b_hi, b_lo, a_hi, a_lo, ...] -> [r_hi, r_lo, q_hi, q_lo ...], where r = a % b, q = a / b
 export.checked_divmod
-    exec.u32assert4
-    exec.unchecked_divmod
+    u32assert.2         # make sure divisor consists of 32-bit limbs
+    adv.u64div          # inject the quotient and the remainder into the advice tape
+
+    push.adv.1          # read the quotient from the advice tape and make sure it consists of
+    push.adv.1          # 32-bit limbs
+    u32assert.2
+
+    dup.3               # multiply quotient by the divisor and make sure the resulting value
+    dup.2               # fits into 2 32-bit limbs
+    u32overflowing_mul
+    dup.4
+    dup.4
+    u32overflowing_madd
+    eq.0
+    assert
+    dup.5
+    dup.3
+    u32overflowing_madd
+    eq.0
+    assert
+    dup.4
+    dup.3
+    mul
+    eq.0
+    assert
+
+    push.adv.1          # read the remainder from the advice tape and make sure it consists of
+    push.adv.1          # 32-bit limbs
+    u32assert.2
+
+    movup.7             # make sure the divisor is greater than the remainder. this also consumes
+    movup.7             # the divisor
+    dup.3
+    dup.3
+    exec.unchecked_gt
+    assert
+
+    dup.1               # add remainder to the previous result
+    movup.4
+    u32overflowing_add
+    movup.4
+    dup.3
+    u32overflowing_add3
+    eq.0
+    assert
+
+    movup.6             # make sure the result we got is equal to the dividend
+    u32assert.2
+    assert_eq
+    movup.5
+    u32assert.2
```

Comment:
I'd also add a comment here to explain that `u32assert.2`s are used to ensure that the dividend consists of two 32-bit limbs (at least that's what I think is happening).